src/network/core/http_winhttp.cpp

```diff
@@ -0,0 +1,301 @@
+		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), buffer, sizeof(buffer), NULL) == 0) {
+		return fmt::format("unknown error {}", error_code);
+	}
+
+	return buffer;
+}
+
+/**
+ * Callback from the WinHTTP library, called when-ever something changes about the HTTP request status.
+ *
+ * The callback needs to call some WinHttp functions for certain states, so WinHttp continues
+ * to read the request. This also allows us to abort when things go wrong, by simply not calling
+ * those functions.
+ * Comments with "Next step:" mark where WinHttp needs a call to continue.
+ *
+ * @param code    The code of the event.
+ * @param info    The information about the event.
+ * @param length  The length of the information.
+ */
+void NetworkHTTPRequest::WinHttpCallback(DWORD code, void *info, DWORD length)
+{
+	switch (code) {
+		case WINHTTP_CALLBACK_STATUS_RESOLVING_NAME:
+		case WINHTTP_CALLBACK_STATUS_NAME_RESOLVED:
+		case WINHTTP_CALLBACK_STATUS_CONNECTING_TO_SERVER:
+		case WINHTTP_CALLBACK_STATUS_CONNECTED_TO_SERVER:
+		case WINHTTP_CALLBACK_STATUS_SENDING_REQUEST:
+		case WINHTTP_CALLBACK_STATUS_REQUEST_SENT:
+		case WINHTTP_CALLBACK_STATUS_RECEIVING_RESPONSE:
+		case WINHTTP_CALLBACK_STATUS_RESPONSE_RECEIVED:
+		case WINHTTP_CALLBACK_STATUS_CLOSING_CONNECTION:
+		case WINHTTP_CALLBACK_STATUS_CONNECTION_CLOSED:
+		case WINHTTP_CALLBACK_STATUS_HANDLE_CREATED:
+		case WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING:
+			/* We don't care about these events, and explicitly ignore them. */
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_REDIRECT:
+			/* Make sure we are not in a redirect loop. */
+			if (this->depth++ > 5) {
+				Debug(net, 0, "HTTP request failed: too many redirects");
+				this->callback->OnFailure();
+				this->finished = true;
+				return;
+			}
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE:
+			/* Next step: read response. */
+			WinHttpReceiveResponse(this->request, nullptr);
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_HEADERS_AVAILABLE:
+		{
+			/* Retrieve the status code. */
+			DWORD status_code = 0;
+			DWORD status_code_size = sizeof(status_code);
+			WinHttpQueryHeaders(this->request, WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER, WINHTTP_HEADER_NAME_BY_INDEX, &status_code, &status_code_size, WINHTTP_NO_HEADER_INDEX);
+			Debug(net, 3, "HTTP request status code: {}", status_code);
+
+			/* If there is any error, we simply abort the request. */
+			if (status_code >= 400) {
+				Debug(net, 0, "HTTP request failed: status-code {}", status_code);
+				this->callback->OnFailure();
+				this->finished = true;
+				return;
+			}
+
+			/* Next step: query for any data. */
+			WinHttpQueryDataAvailable(this->request, nullptr);
+		} break;
+
+		case WINHTTP_CALLBACK_STATUS_DATA_AVAILABLE:
+		{
+			/* Retrieve the amount of data available to process. */
+			DWORD size = *(DWORD *)info;
+
+			/* Next step: read the data in a temporary buffer. */
+			char *buffer = size == 0 ? nullptr : MallocT<char>(size);
```

Comment:
If the data is available, then it should be in some sort of buffer. Why is this copy of the data needed, especially if we're copying it later to disk as well?

Reply:
That is a question for the WinHttp API gods, and not one I can answer :( This is how the documentation says how to implement this. Not sure what more I can add, sorry :)

Reply:
To quote, on the same page: `Although calling WinHttpReadData immediately after completion of WinHttpReceiveResponse avoids the expense of a buffer copy, doing so requires that the application use a fixed-length buffer for reading.`
So it is possible to do it without the extra copy, but I have no clue what the impact of it would be. It's mostly a check question to prevent the extra copying if that's reasonably easy to attain.

Reply:
How I read that, is that it internally doesn't allocate an additional buffer. So out of our hands :)

So yeah, let's do this for now; hopefully someone with more knowledge about WinHttp comes along and goes: NOOOO, and fixes it :D